tests/components/sensor/test_canary.py

```diff
@@ -0,0 +1,134 @@
+"""The tests for the Canary sensor platform."""
+import copy
+import unittest
+from unittest.mock import patch, Mock, MagicMock, PropertyMock
+
+from canary.api import SensorType, Reading
+from homeassistant.components import canary as base_canary
+from homeassistant.components.canary import DATA_CANARY
+from homeassistant.components.sensor import canary
+from homeassistant.components.sensor.canary import CanarySensor
+from tests.common import (get_test_home_assistant)
+from tests.components.test_canary import API_LOCATIONS
+
+VALID_CONFIG = {
+    "canary": {
+        "username": "foo@bar.org",
+        "password": "bar",
+    }
+}
+
+
+def mock_device(device_id, name):
+    device = MagicMock()
+    type(device).device_id = PropertyMock(return_value=device_id)
+    type(device).name = PropertyMock(return_value=name)
+    return device
+
+
+def mock_location(name, is_celsius=True):
+    location = MagicMock()
+    type(location).name = PropertyMock(return_value=name)
+    type(location).is_celsius = PropertyMock(return_value=is_celsius)
+    return location
+
+
+class TestCanarySensorSetup(unittest.TestCase):
+    """Test the Canary platform."""
+
+    DEVICES = []
+
+    def add_devices(self, devices, action):
+        """Mock add devices."""
+        for device in devices:
+            self.DEVICES.append(device)
+
+    def setUp(self):
+        """Initialize values for this testcase class."""
+        self.hass = get_test_home_assistant()
+        self.config = copy.deepcopy(VALID_CONFIG)
+
+    def tearDown(self):
+        """Stop everything that was started."""
+        self.hass.stop()
+
+    @patch('homeassistant.components.canary.CanaryData')
+    def test_setup_sensors(self, mock_canary):
+        """Test the Canary senskor class and methods."""
+        base_canary.setup(self.hass, self.config)
+
+        self.hass.data[DATA_CANARY] = mock_canary()
+        self.hass.data[DATA_CANARY].locations = API_LOCATIONS
+
+        canary.setup_platform(self.hass, self.config, self.add_devices, None)
+
+        self.assertEqual(6, len(self.DEVICES))
+
+    def test_celsius_temperature_sensor(self):
+        device = mock_device(10, "Family Room")
+        location = mock_location("Home", True)
+
+        data = Mock()
+        data.get_readings.return_value = [Reading({
```

Comment:
I think you should mock the api classes/instances instead, so we don't need to add the library to the test requirements.

Reply:
I got rid of api classes as much as I can. But SensorType (api class) which is used in sensor.canary.Canary.unit_of_measurement. I'm using it to determine the unit.

If I mock sensor type, then the equality will not work. Do you have any suggestion? I can change  code in unit_of_measurement to something like:

```
if self._sensor_type.value == SensorType.TEMPERATURE.value:
```

But do you think that's good?

Reply:
Can't you patch the return value of the mock sensor type to be an enum too?

Reply:
But that'd require importing API class (SensorType) in the test. I thought that's something you want to avoid?

Reply:
No, it should be enough to make sure that the enum attribute is the same on the mock object.

Reply:
Is this what you mean?

```
        with patch('canary.api.SensorType') as mock_sensor_type:
            type(mock_sensor_type).value = PropertyMock(return_value="temperature")

            data = Mock()
            data.get_readings.return_value = [
                mock_reading(mock_sensor_type, 21.1234)]

            sensor = CanarySensor(data, mock_sensor_type, location, device)
            sensor.update()

            self.assertEqual("Home Family Room Temperature", sensor.name)
            self.assertEqual("sensor_canary_10_temperature", sensor.unique_id)
            self.assertEqual("°C", sensor.unit_of_measurement)
```

I'm getting "AssertionError: '°C' != None" which is expected because unit_of_measurement is implemented like this:

```
    @property
    def unit_of_measurement(self):
        """Return the unit of measurement this sensor expresses itself in."""
        from canary.api import SensorType
        if self._sensor_type == SensorType.TEMPERATURE:
            return TEMP_CELSIUS if self._is_celsius else TEMP_FAHRENHEIT
        elif self._sensor_type == SensorType.HUMIDITY:
            return "%"
        elif self._sensor_type == SensorType.AIR_QUALITY:
            return ""
        return None
```

The part I don't know how to do is, how do you mock the enum and still pass the equality check?

Reply:
I think I was wrong. It's not possible to mock identity. Ok, let's keep the library as a requirement, so we don't waste more time on this. Sorry for the hassle.

Reply:
Ok good to know :)

Is there anything else needed for this PR? I’ll work on the doc shortly